<!DOCTYPE html>
<!-- saved from url=(0093)file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">


<meta name="generator" content="pandoc">
<meta http-equiv="X-UA-Compatible" content="IE=EDGE">

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Example run of CoExpansionValidation</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      code.sourceCode > span { display: inline-block; line-height: 1.25; }
  code.sourceCode > span { color: inherit; text-decoration: inherit; }
  code.sourceCode > span:empty { height: 1.2em; }
  .sourceCode { overflow: visible; }
  code.sourceCode { white-space: pre; position: relative; }
  div.sourceCode { margin: 1em 0; }
  pre.sourceCode { margin: 0; }
  @media screen {
  div.sourceCode { overflow: auto; }
  }
  @media print {
  code.sourceCode { white-space: pre-wrap; }
  code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
  }
  pre.numberSource code
    { counter-reset: source-line 0; }
  pre.numberSource code > span
    { position: relative; left: -4em; counter-increment: source-line; }
  pre.numberSource code > span > a:first-child::before
    { content: counter(source-line);
      position: relative; left: -1em; text-align: right; vertical-align: baseline;
      border: none; display: inline-block;
      -webkit-touch-callout: none; -webkit-user-select: none;
      -khtml-user-select: none; -moz-user-select: none;
      -ms-user-select: none; user-select: none;
      padding: 0 4px; width: 4em;
      color: #aaaaaa;
    }
  pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
  div.sourceCode
    {   }
  @media screen {
  code.sourceCode > span > a:first-child::before { text-decoration: underline; }
  }
  code span.al { color: #ff0000; font-weight: bold; } /* Alert */
  code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
  code span.at { color: #7d9029; } /* Attribute */
  code span.bn { color: #40a070; } /* BaseN */
  code span.bu { } /* BuiltIn */
  code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
  code span.ch { color: #4070a0; } /* Char */
  code span.cn { color: #880000; } /* Constant */
  code span.co { color: #60a0b0; font-style: italic; } /* Comment */
  code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
  code span.do { color: #ba2121; font-style: italic; } /* Documentation */
  code span.dt { color: #902000; } /* DataType */
  code span.dv { color: #40a070; } /* DecVal */
  code span.er { color: #ff0000; font-weight: bold; } /* Error */
  code span.ex { } /* Extension */
  code span.fl { color: #40a070; } /* Float */
  code span.fu { color: #06287e; } /* Function */
  code span.im { } /* Import */
  code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  code span.kw { color: #007020; font-weight: bold; } /* Keyword */
  code span.op { color: #666666; } /* Operator */
  code span.ot { color: #007020; } /* Other */
  code span.pp { color: #bc7a00; } /* Preprocessor */
  code span.sc { color: #4070a0; } /* SpecialChar */
  code span.ss { color: #bb6688; } /* SpecialString */
  code span.st { color: #4070a0; } /* String */
  code span.va { color: #19177c; } /* Variable */
  code span.vs { color: #4070a0; } /* VerbatimString */
  code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




<script type="text/javascript" src="./Example run of CoExpansionValidation_files/MathJax.js.下载"></script><style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Menu_Button .MathJax_Hover_Arrow {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 4px; -webkit-border-radius: 4px; -moz-border-radius: 4px; -khtml-border-radius: 4px; font-family: 'Courier New',Courier; font-size: 9px; color: #F0F0F0}
.MathJax_Menu_Button .MathJax_Hover_Arrow span {display: block; background-color: #AAA; border: 1px solid; border-radius: 3px; line-height: 0; padding: 4px}
.MathJax_Hover_Arrow:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_Hover_Arrow:hover span {background-color: #CCC!important}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_About.MathJax_MousePost {outline: none}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 2px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 2px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; padding-top: .25em; color: #666666; font-size: .75em}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: 1em}
.MathJax_MenuRadioCheck.RTL {right: 1em; left: auto}
.MathJax_MenuLabel {padding: 2px 2em 4px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #CCCCCC; margin: 4px 1px 0px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: Highlight; color: HighlightText}
.MathJax_MenuDisabled:focus, .MathJax_MenuLabel:focus {background-color: #E8E8E8}
.MathJax_ContextMenu:focus {outline: none}
.MathJax_ContextMenu .MathJax_MenuItem:focus {outline: none}
#MathJax_AboutClose {top: .2em; right: .2em}
.MathJax_Menu .MathJax_MenuClose {top: -10px; left: -10px}
.MathJax_MenuClose {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; font-family: 'Courier New',Courier; font-size: 24px; color: #F0F0F0}
.MathJax_MenuClose span {display: block; background-color: #AAA; border: 1.5px solid; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; line-height: 0; padding: 8px 0 6px}
.MathJax_MenuClose:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_MenuClose:hover span {background-color: #CCC!important}
.MathJax_MenuClose:hover:focus {outline: none}
</style><style type="text/css">.MathJax_Preview .MJXf-math {color: inherit!important}
</style><style type="text/css">.MJX_Assistive_MathML {position: absolute!important; top: 0; left: 0; clip: rect(1px, 1px, 1px, 1px); padding: 1px 0 0 0!important; border: 0!important; height: 1px!important; width: 1px!important; overflow: hidden!important; display: block!important; -webkit-touch-callout: none; -webkit-user-select: none; -khtml-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none}
.MJX_Assistive_MathML.MJX_Assistive_MathML_Block {width: 100%!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1em; bottom: 1.5em; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style><style type="text/css">.MJXp-script {font-size: .8em}
.MJXp-right {-webkit-transform-origin: right; -moz-transform-origin: right; -ms-transform-origin: right; -o-transform-origin: right; transform-origin: right}
.MJXp-bold {font-weight: bold}
.MJXp-italic {font-style: italic}
.MJXp-scr {font-family: MathJax_Script,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-frak {font-family: MathJax_Fraktur,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-sf {font-family: MathJax_SansSerif,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-cal {font-family: MathJax_Caligraphic,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-mono {font-family: MathJax_Typewriter,'Times New Roman',Times,STIXGeneral,serif}
.MJXp-largeop {font-size: 150%}
.MJXp-largeop.MJXp-int {vertical-align: -.2em}
.MJXp-math {display: inline-block; line-height: 1.2; text-indent: 0; font-family: 'Times New Roman',Times,STIXGeneral,serif; white-space: nowrap; border-collapse: collapse}
.MJXp-display {display: block; text-align: center; margin: 1em 0}
.MJXp-math span {display: inline-block}
.MJXp-box {display: block!important; text-align: center}
.MJXp-box:after {content: " "}
.MJXp-rule {display: block!important; margin-top: .1em}
.MJXp-char {display: block!important}
.MJXp-mo {margin: 0 .15em}
.MJXp-mfrac {margin: 0 .125em; vertical-align: .25em}
.MJXp-denom {display: inline-table!important; width: 100%}
.MJXp-denom > * {display: table-row!important}
.MJXp-surd {vertical-align: top}
.MJXp-surd > * {display: block!important}
.MJXp-script-box > *  {display: table!important; height: 50%}
.MJXp-script-box > * > * {display: table-cell!important; vertical-align: top}
.MJXp-script-box > *:last-child > * {vertical-align: bottom}
.MJXp-script-box > * > * > * {display: block!important}
.MJXp-mphantom {visibility: hidden}
.MJXp-munderover {display: inline-table!important}
.MJXp-over {display: inline-block!important; text-align: center}
.MJXp-over > * {display: block!important}
.MJXp-munderover > * {display: table-row!important}
.MJXp-mtable {vertical-align: .25em; margin: 0 .125em}
.MJXp-mtable > * {display: inline-table!important; vertical-align: middle}
.MJXp-mtr {display: table-row!important}
.MJXp-mtd {display: table-cell!important; text-align: center; padding: .5em 0 0 .5em}
.MJXp-mtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-mlabeledtr {display: table-row!important}
.MJXp-mlabeledtr > .MJXp-mtd:first-child {padding-left: 0}
.MJXp-mlabeledtr:first-child > .MJXp-mtd {padding-top: 0}
.MJXp-merror {background-color: #FFFF88; color: #CC0000; border: 1px solid #CC0000; padding: 1px 3px; font-style: normal; font-size: 90%}
.MJXp-scale0 {-webkit-transform: scaleX(.0); -moz-transform: scaleX(.0); -ms-transform: scaleX(.0); -o-transform: scaleX(.0); transform: scaleX(.0)}
.MJXp-scale1 {-webkit-transform: scaleX(.1); -moz-transform: scaleX(.1); -ms-transform: scaleX(.1); -o-transform: scaleX(.1); transform: scaleX(.1)}
.MJXp-scale2 {-webkit-transform: scaleX(.2); -moz-transform: scaleX(.2); -ms-transform: scaleX(.2); -o-transform: scaleX(.2); transform: scaleX(.2)}
.MJXp-scale3 {-webkit-transform: scaleX(.3); -moz-transform: scaleX(.3); -ms-transform: scaleX(.3); -o-transform: scaleX(.3); transform: scaleX(.3)}
.MJXp-scale4 {-webkit-transform: scaleX(.4); -moz-transform: scaleX(.4); -ms-transform: scaleX(.4); -o-transform: scaleX(.4); transform: scaleX(.4)}
.MJXp-scale5 {-webkit-transform: scaleX(.5); -moz-transform: scaleX(.5); -ms-transform: scaleX(.5); -o-transform: scaleX(.5); transform: scaleX(.5)}
.MJXp-scale6 {-webkit-transform: scaleX(.6); -moz-transform: scaleX(.6); -ms-transform: scaleX(.6); -o-transform: scaleX(.6); transform: scaleX(.6)}
.MJXp-scale7 {-webkit-transform: scaleX(.7); -moz-transform: scaleX(.7); -ms-transform: scaleX(.7); -o-transform: scaleX(.7); transform: scaleX(.7)}
.MJXp-scale8 {-webkit-transform: scaleX(.8); -moz-transform: scaleX(.8); -ms-transform: scaleX(.8); -o-transform: scaleX(.8); transform: scaleX(.8)}
.MJXp-scale9 {-webkit-transform: scaleX(.9); -moz-transform: scaleX(.9); -ms-transform: scaleX(.9); -o-transform: scaleX(.9); transform: scaleX(.9)}
.MathJax_PHTML .noError {vertical-align: ; font-size: 90%; text-align: left; color: black; padding: 1px 3px; border: 1px solid}
</style></head>

<body><div id="MathJax_Message" style="display: none;"></div>




<h1 class="title toc-ignore">Example run of CoExpansionValidation</h1>



<p>This is a markdown file showing how to use the CoExpansionValidation R package to testing the power of genetic datasets for detecting co-expansion events across species.</p>
<p>Tests of synchronous community expansion are commonly conducted using hierarchical Approximate Bayesian Computation (hABC), a statistical framework for inferring the degree of concordance across species. However, this framework is often used without demonstrating adequate performance.</p>
<p>This file will go through step-by-step an example of how to generate one pseudo-observed dataset, and use ABC to “infer” the number of co-expansion events. Obviously, to apply these performance assessment in a empirical study, many replicates of pseudo-observed datasets are needed.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb1-1" aria-hidden="true"></a><span class="kw">library</span>(CoExpansionValidation)</span>
<span id="cb1-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb1-2" aria-hidden="true"></a><span class="co">#library(devtools)</span></span>
<span id="cb1-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb1-3" aria-hidden="true"></a><span class="co">#load_all()</span></span>
<span id="cb1-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb1-4" aria-hidden="true"></a><span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> F)</span></code></pre></div>
<div id="pre-requisite" class="section level3">
<h3><strong>Pre-requisite</strong></h3>
<p>external programs that need to be installed before running:</p>
<ul>
<li><a href="https://www.python.org/">Python 2.x</a> &gt;= 2.4<br>
</li>
<li><a href="http://www.stanford.edu/group/hadlylab/ssc/">BayeSSC - Serial Simcoal</a><br>
</li>
<li><a href="https://github.com/UH-Bioinformatics/hBayeSSC">hBayeSSC.py</a><br>
</li>
<li>[msReject]</li>
</ul>
</div>
<div id="generating-a-test-dataset" class="section level1">
<h1><strong>1.Generating A Test Dataset</strong></h1>
<div id="generate-divergent-time-for-species" class="section level2">
<h2>1) Generate divergent time for species</h2>
<div id="i-determine-the-number-of-co-expansion-events" class="section level3">
<h3>i) Determine the number of co-expansion events</h3>
<p>Two options:</p>
<ul>
<li>user specifies it along with the number of species and assign species to the events randomly or evenly:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-1" aria-hidden="true"></a>species&lt;-<span class="dv">1</span><span class="op">:</span><span class="dv">10</span> <span class="co"># or a vector of species names</span></span>
<span id="cb2-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-2" aria-hidden="true"></a>coevents&lt;-<span class="dv">2</span></span>
<span id="cb2-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-3" aria-hidden="true"></a><span class="co">#assigning species randomly</span></span>
<span id="cb2-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-4" aria-hidden="true"></a>species.assignment&lt;-<span class="kw">assign_species_to_events</span>(<span class="dt">species =</span> species,<span class="dt">nco.events =</span> coevents,<span class="dt">even =</span> F)</span>
<span id="cb2-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-5" aria-hidden="true"></a>species.assignment</span>
<span id="cb2-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-6" aria-hidden="true"></a><span class="co">#&gt; [[1]]</span></span>
<span id="cb2-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-7" aria-hidden="true"></a><span class="co">#&gt; [1]  1  2  3  4  7  8  9 10</span></span>
<span id="cb2-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-8" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb2-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-9" aria-hidden="true"></a><span class="co">#&gt; [[2]]</span></span>
<span id="cb2-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-10" aria-hidden="true"></a><span class="co">#&gt; [1] 5 6</span></span>
<span id="cb2-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-11" aria-hidden="true"></a><span class="co">#assigning species evenly</span></span>
<span id="cb2-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-12" aria-hidden="true"></a>species.assignment&lt;-<span class="kw">assign_species_to_events</span>(<span class="dt">species =</span> species,<span class="dt">nco.events =</span> coevents,<span class="dt">even =</span> T)</span>
<span id="cb2-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-13" aria-hidden="true"></a>species.assignment</span>
<span id="cb2-14"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-14" aria-hidden="true"></a><span class="co">#&gt; [[1]]</span></span>
<span id="cb2-15"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-15" aria-hidden="true"></a><span class="co">#&gt; [1] 1 2 3 4 5</span></span>
<span id="cb2-16"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-16" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb2-17"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-17" aria-hidden="true"></a><span class="co">#&gt; [[2]]</span></span>
<span id="cb2-18"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb2-18" aria-hidden="true"></a><span class="co">#&gt; [1]  6  7  8  9 10</span></span></code></pre></div>
<ul>
<li>user specifies the number of species and the <em>alpha</em> for Dirichlet process, and simulate the number of co-expansion events</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-1" aria-hidden="true"></a>species&lt;-<span class="dv">1</span><span class="op">:</span><span class="dv">10</span></span>
<span id="cb3-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-2" aria-hidden="true"></a>alpha&lt;-<span class="fl">1.1</span></span>
<span id="cb3-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-3" aria-hidden="true"></a>species.assignment&lt;-<span class="kw">generate_coevent_number_dirichlet</span>(<span class="dt">species =</span> species,<span class="dt">alpha =</span> alpha,<span class="dt">maxevent=</span><span class="dv">4</span>)</span>
<span id="cb3-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-4" aria-hidden="true"></a>species.assignment</span>
<span id="cb3-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-5" aria-hidden="true"></a><span class="co">#&gt; [[1]]</span></span>
<span id="cb3-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-6" aria-hidden="true"></a><span class="co">#&gt; [1]  1  4  5  6  8  9 10</span></span>
<span id="cb3-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-7" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb3-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-8" aria-hidden="true"></a><span class="co">#&gt; [[2]]</span></span>
<span id="cb3-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-9" aria-hidden="true"></a><span class="co">#&gt; [1] 2</span></span>
<span id="cb3-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-10" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb3-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-11" aria-hidden="true"></a><span class="co">#&gt; [[3]]</span></span>
<span id="cb3-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-12" aria-hidden="true"></a><span class="co">#&gt; [1] 3 7</span></span>
<span id="cb3-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb3-13" aria-hidden="true"></a>coevents&lt;-<span class="kw">length</span>(species.assignment)</span></code></pre></div>
</div>
<div id="ii-generating-the-expansion-time-for-coexpansion-events" class="section level3">
<h3>ii) Generating the expansion time for coexpansion events</h3>
<p>User can specify the time for each event (time from present), or randomly sample from a time range. Buffer can be added around a event, so co-expansion events are at least certain generations away from each other. No required unit for time and buffer–below we are going to specify the generation time to turn them into number of generations.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-1" aria-hidden="true"></a>exp.time&lt;-<span class="kw">c</span>(<span class="dv">30000</span>,<span class="dv">50000</span>) </span>
<span id="cb4-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-2" aria-hidden="true"></a><span class="co"># if specifying the time, make sure its length equals </span></span>
<span id="cb4-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-3" aria-hidden="true"></a><span class="co"># the coevents specified above</span></span>
<span id="cb4-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-4" aria-hidden="true"></a></span>
<span id="cb4-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-5" aria-hidden="true"></a>time.range&lt;-<span class="kw">c</span>(<span class="dv">30000</span>,<span class="dv">50000</span>)</span>
<span id="cb4-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-6" aria-hidden="true"></a>buffer&lt;-<span class="dv">5000</span></span>
<span id="cb4-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-7" aria-hidden="true"></a>exp.time&lt;-<span class="kw">generate_cotime_with_buffer</span>(<span class="dt">time.range =</span> time.range,<span class="dt">nco.events =</span>coevents,<span class="dt">buffer =</span>  buffer)</span>
<span id="cb4-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-8" aria-hidden="true"></a>exp.time</span>
<span id="cb4-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb4-9" aria-hidden="true"></a><span class="co">#&gt; [1] 37248.81 43284.50 31073.80</span></span></code></pre></div>
<p>User can get a vector of the generated time for each species, easier to write out to a file</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-1" aria-hidden="true"></a></span>
<span id="cb5-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-2" aria-hidden="true"></a>x&lt;-<span class="kw">species_exp_time</span>(species.assignment,exp.time)</span>
<span id="cb5-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-3" aria-hidden="true"></a>x</span>
<span id="cb5-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-4" aria-hidden="true"></a><span class="co">#&gt;        1       10        2        3        4        5        6        7 </span></span>
<span id="cb5-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-5" aria-hidden="true"></a><span class="co">#&gt; 37248.81 37248.81 43284.50 31073.80 37248.81 37248.81 37248.81 31073.80 </span></span>
<span id="cb5-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-6" aria-hidden="true"></a><span class="co">#&gt;        8        9 </span></span>
<span id="cb5-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb5-7" aria-hidden="true"></a><span class="co">#&gt; 37248.81 37248.81</span></span></code></pre></div>
</div>
<div id="iii-simulation-with-bayessc" class="section level3">
<h3>iii) Simulation with BayeSSC</h3>
<p>Here, we provide a wrapper function for running BayeSCC from R. BayeSSC itself has very detail instructions on <a href="https://web.stanford.edu/group/hadlylab/ssc/index.html#IO">their webpage</a>. User need to give:</p>
<ul>
<li>path to the BayeSCC executable (absolute path or relative to current working directory; including the executable name)</li>
</ul>
<p>And then, there are two options:</p>
<ol style="list-style-type: lower-alpha">
<li>a configuration file: a table (TAB delimited file) consists of columns with following header names</li>
</ol>
<table>
<colgroup>
<col width="55%">
<col width="44%">
</colgroup>
<thead>
<tr class="header">
<th><strong>Column name</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>species<sup>1</sup></td>
<td>Name of taxa</td>
</tr>
<tr class="even">
<td>locinum</td>
<td>Number of loci</td>
</tr>
<tr class="odd">
<td>nsam</td>
<td>Number of haplotypes to be simulated</td>
</tr>
<tr class="even">
<td>nsites</td>
<td>Number of base pairs</td>
</tr>
<tr class="odd">
<td>tstv</td>
<td>% transitions</td>
</tr>
<tr class="even">
<td>gamma</td>
<td>Gamma distribution for substitution rate heterogeneity, two numbers separated by space</td>
</tr>
<tr class="odd">
<td>mutation rate<sup>2</sup></td>
<td>The locus mutation rate per generation</td>
</tr>
<tr class="even">
<td>gen<sup>3</sup></td>
<td>The generation length</td>
</tr>
<tr class="odd">
<td>Ne</td>
<td>The Effective population size</td>
</tr>
<tr class="even">
<td>popratio</td>
<td>The ratio between ancestral and current population size, &lt;1 for expansion and &gt;1 for shrinkage</td>
</tr>
<tr class="odd">
<td>eventtime.generation<sup>4</sup></td>
<td>The time of the historical event in unit of generation(optional)</td>
</tr>
</tbody>
</table>
<p><sup>1</sup>:species can appear in multiple lines if it has multiple types of loci; or if user prefer to use one row for one locus at a time. However, all loci in one species share the same generation time, population size and expansion/shrinkage history, we will only use data of these columns from the first row of a species.<br>
<sup>2</sup>: mutation rate (and population size, and other number columns) can take BayeSCC-style prior distributions. For example, {U:1,299} for uniform distribution between 1 and 299. see <a href="https://web.stanford.edu/group/hadlylab/ssc/index.html#IO">BayeSCC webpage</a> for details.<br>
<sup>3</sup>:generation length need to be in the same unit as the time range and buffer provided before. if previously used MYA, then here would need the generation time in unit of million years.<br>
<sup>4</sup>:optional, if not provided, will be calculated automatically with generation time and previously simulated event time</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-1" aria-hidden="true"></a>path_to_bayessc&lt;-<span class="st">"BayeSSC.exe"</span> <span class="co">#windows version, linux or mac starts with ./</span></span>
<span id="cb6-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-2" aria-hidden="true"></a></span>
<span id="cb6-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-3" aria-hidden="true"></a><span class="co">#check the conf data frame --an example of the configuration table-- included in the package</span></span>
<span id="cb6-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-4" aria-hidden="true"></a><span class="kw">head</span>(conf)</span>
<span id="cb6-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-5" aria-hidden="true"></a><span class="co">#&gt;   species locinum nsam nsite tstv  gamma gen            Ne popratio</span></span>
<span id="cb6-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-6" aria-hidden="true"></a><span class="co">#&gt; 1       1       5   40   600 0.33 0.5  4   2   {U:200,300}     0.01</span></span>
<span id="cb6-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-7" aria-hidden="true"></a><span class="co">#&gt; 2       1       6   20   100 0.80 1.8 10   2   {U:200,300}     0.01</span></span>
<span id="cb6-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-8" aria-hidden="true"></a><span class="co">#&gt; 3       2       1   20   500 0.90 1.8 10   1   {U:800,900}     0.05</span></span>
<span id="cb6-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-9" aria-hidden="true"></a><span class="co">#&gt; 4       3       1   20   500 0.90 1.8 10   1 {U:800,20000}     0.05</span></span>
<span id="cb6-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-10" aria-hidden="true"></a><span class="co">#&gt; 5       4       1   40   500 0.90 1.8 10   1  {U:800,9000}     0.03</span></span>
<span id="cb6-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-11" aria-hidden="true"></a><span class="co">#&gt; 6       5       8   20   500 0.60 1.8 10   1   {U:800,900}     0.05</span></span>
<span id="cb6-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-12" aria-hidden="true"></a><span class="co">#&gt;         mutationrate</span></span>
<span id="cb6-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-13" aria-hidden="true"></a><span class="co">#&gt; 1 {N:0.0006,0.00002}</span></span>
<span id="cb6-14"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-14" aria-hidden="true"></a><span class="co">#&gt; 2 {N:0.0006,0.00002}</span></span>
<span id="cb6-15"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-15" aria-hidden="true"></a><span class="co">#&gt; 3 {N:0.0006,0.00002}</span></span>
<span id="cb6-16"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-16" aria-hidden="true"></a><span class="co">#&gt; 4 {N:0.0006,0.00002}</span></span>
<span id="cb6-17"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-17" aria-hidden="true"></a><span class="co">#&gt; 5 {N:0.0006,0.00002}</span></span>
<span id="cb6-18"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-18" aria-hidden="true"></a><span class="co">#&gt; 6 {N:0.0006,0.00002}</span></span>
<span id="cb6-19"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-19" aria-hidden="true"></a></span>
<span id="cb6-20"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-20" aria-hidden="true"></a><span class="co">#read your own configuration file:</span></span>
<span id="cb6-21"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-21" aria-hidden="true"></a><span class="co">#configurefile&lt;-"test.conf"</span></span>
<span id="cb6-22"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-22" aria-hidden="true"></a><span class="co">#conf&lt;-read.table(configurefile,header=T,sep="\t",stringsAsFactors = F)</span></span>
<span id="cb6-23"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-23" aria-hidden="true"></a></span>
<span id="cb6-24"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-24" aria-hidden="true"></a><span class="co">#code below will simulate summary statistics with BayeSSC:</span></span>
<span id="cb6-25"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-25" aria-hidden="true"></a><span class="co">#simulatedobs&lt;-runbayeSSC_with_conf(BayeSSCallocation = path_to_bayessc,conf = conf,prefix = 'temp',species.assignment = species.assignment,exp.time = exp.time,intern=F)</span></span>
<span id="cb6-26"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-26" aria-hidden="true"></a></span>
<span id="cb6-27"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-27" aria-hidden="true"></a><span class="co">#the package comes with a simulatedobs data frame, user can check how it looks like</span></span>
<span id="cb6-28"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-28" aria-hidden="true"></a><span class="kw">colnames</span>(simulatedobs)</span>
<span id="cb6-29"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-29" aria-hidden="true"></a><span class="co">#&gt;  [1] "species"   "nsam"      "nsites"    "tstv"      "gamma"     "gen"      </span></span>
<span id="cb6-30"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-30" aria-hidden="true"></a><span class="co">#&gt;  [7] "locuslow"  "locushigh" "Nelow"     "Nehigh"    "SegSites"  "Nucdiv"   </span></span>
<span id="cb6-31"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-31" aria-hidden="true"></a><span class="co">#&gt; [13] "Haptypes"  "HapDiver"  "Pairdiffs" "TajimasD"  "F*"</span></span>
<span id="cb6-32"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-32" aria-hidden="true"></a><span class="kw">head</span>(simulatedobs[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</span>
<span id="cb6-33"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-33" aria-hidden="true"></a><span class="co">#&gt;      species    nsam nsites tstv   gamma</span></span>
<span id="cb6-34"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-34" aria-hidden="true"></a><span class="co">#&gt; [1,] "sp1loci1" "40" "800"  "0.33" "0"  </span></span>
<span id="cb6-35"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-35" aria-hidden="true"></a><span class="co">#&gt; [2,] "sp1loci2" "40" "800"  "0.33" "0"  </span></span>
<span id="cb6-36"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-36" aria-hidden="true"></a><span class="co">#&gt; [3,] "sp1loci3" "40" "800"  "0.33" "0"  </span></span>
<span id="cb6-37"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-37" aria-hidden="true"></a><span class="co">#&gt; [4,] "sp1loci4" "40" "800"  "0.33" "0"  </span></span>
<span id="cb6-38"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-38" aria-hidden="true"></a><span class="co">#&gt; [5,] "sp1loci5" "40" "800"  "0.33" "0"  </span></span>
<span id="cb6-39"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb6-39" aria-hidden="true"></a><span class="co">#&gt; [6,] "sp1loci6" "40" "800"  "0.33" "0"</span></span></code></pre></div>
<ol start="2" style="list-style-type: lower-alpha">
<li>if all loci in all species can use the same .par file for BayeSCC (all loci in a species are of the same type)</li>
</ol>
<ul>
<li>.par file for BayeSCC. Here we only model one expansion event per species so its format is fixed, do not add or delete line from the template file. For the line under “1 historical event”, the time of the event can set to be any integer; the wrapper function will swap in the generated (or set) expansion time from above. No migration; note that “the number of loci” is actually the length in bp for DNA. Users provide other parameters– population size, sample size and etc.<br>
</li>
<li>the number of loci need to be generated per species</li>
<li>generation time (if not set, default to 1)</li>
</ul>
<p>Below is the template file, already scanned in as part of the package data, BaySSCtemplate.par</p>
<blockquote>
<p>BayeSSC//Number of population samples 1<br>
//Population sizes<br>
{U:500000,500000}<br>
//Sample sizes<br>
40<br>
//Growth rates<br>
0<br>
//Number of migration matrices : If 0 : No migration between demes<br>
0<br>
//Historical event format:<br>
1 historical event<br>
399137 0 0 1 0.01 0 0<br>
//mutation rate<br>
{U:0.00008,0.00008}<br>
//Number of independent loci<br>
800<br>
//Data type, tvts<br>
DNA 0.33<br>
//Gamma distribution for mutation<br>
0 0</p>
</blockquote>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-1" aria-hidden="true"></a>path_to_bayessc&lt;-<span class="st">"BayeSSC.exe"</span> <span class="co">#windows version, linux or mac starts with ./</span></span>
<span id="cb7-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-2" aria-hidden="true"></a></span>
<span id="cb7-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-3" aria-hidden="true"></a><span class="co">#show the templateparfile with this package</span></span>
<span id="cb7-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-4" aria-hidden="true"></a>templateparfile</span>
<span id="cb7-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-5" aria-hidden="true"></a><span class="co">#&gt;  [1] "BayeSSC//Number of population samples - tmp"                       </span></span>
<span id="cb7-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-6" aria-hidden="true"></a><span class="co">#&gt;  [2] "1"                                                                 </span></span>
<span id="cb7-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-7" aria-hidden="true"></a><span class="co">#&gt;  [3] "//Population sizes"                                                </span></span>
<span id="cb7-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-8" aria-hidden="true"></a><span class="co">#&gt;  [4] "{U:500000,500000}"                                                 </span></span>
<span id="cb7-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-9" aria-hidden="true"></a><span class="co">#&gt;  [5] "//Sample sizes"                                                    </span></span>
<span id="cb7-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-10" aria-hidden="true"></a><span class="co">#&gt;  [6] "40"                                                                </span></span>
<span id="cb7-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-11" aria-hidden="true"></a><span class="co">#&gt;  [7] "//Growth rates"                                                    </span></span>
<span id="cb7-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-12" aria-hidden="true"></a><span class="co">#&gt;  [8] "0"                                                                 </span></span>
<span id="cb7-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-13" aria-hidden="true"></a><span class="co">#&gt;  [9] "//Number of migration matrices : If 0 : No migration between demes"</span></span>
<span id="cb7-14"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-14" aria-hidden="true"></a><span class="co">#&gt; [10] "0"                                                                 </span></span>
<span id="cb7-15"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-15" aria-hidden="true"></a><span class="co">#&gt; [11] "//Historical event format:"                                        </span></span>
<span id="cb7-16"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-16" aria-hidden="true"></a><span class="co">#&gt; [12] "1 historical event"                                                </span></span>
<span id="cb7-17"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-17" aria-hidden="true"></a><span class="co">#&gt; [13] "399137 0 0 1 0.01 0 0"                                             </span></span>
<span id="cb7-18"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-18" aria-hidden="true"></a><span class="co">#&gt; [14] "//mutation rate"                                                   </span></span>
<span id="cb7-19"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-19" aria-hidden="true"></a><span class="co">#&gt; [15] "{U:0.00008,0.00008}"                                               </span></span>
<span id="cb7-20"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-20" aria-hidden="true"></a><span class="co">#&gt; [16] "//Number of independent loci"                                      </span></span>
<span id="cb7-21"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-21" aria-hidden="true"></a><span class="co">#&gt; [17] "800"                                                               </span></span>
<span id="cb7-22"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-22" aria-hidden="true"></a><span class="co">#&gt; [18] "//Data type, num loci, rec.rate, mut rate, gamma, shape"           </span></span>
<span id="cb7-23"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-23" aria-hidden="true"></a><span class="co">#&gt; [19] "DNA 0.33"                                                          </span></span>
<span id="cb7-24"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-24" aria-hidden="true"></a><span class="co">#&gt; [20] "//Gamma distribution for mutation"                                 </span></span>
<span id="cb7-25"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-25" aria-hidden="true"></a><span class="co">#&gt; [21] "0 0"</span></span>
<span id="cb7-26"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-26" aria-hidden="true"></a></span>
<span id="cb7-27"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-27" aria-hidden="true"></a><span class="co"># you can write this template out to a file and edit it with any text editor</span></span>
<span id="cb7-28"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-28" aria-hidden="true"></a></span>
<span id="cb7-29"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-29" aria-hidden="true"></a><span class="kw">cat</span>(<span class="kw">paste</span>(templateparfile,<span class="dt">sep =</span> <span class="st">''</span>,<span class="dt">collapse =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>),<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>,<span class="dt">sep=</span><span class="st">''</span>,<span class="dt">file=</span><span class="st">"test.par"</span>)</span>
<span id="cb7-30"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-30" aria-hidden="true"></a></span>
<span id="cb7-31"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-31" aria-hidden="true"></a>bayessc_par_file&lt;-<span class="st">'test.par'</span></span>
<span id="cb7-32"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-32" aria-hidden="true"></a>nloci&lt;-<span class="dv">10</span></span>
<span id="cb7-33"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-33" aria-hidden="true"></a>gen&lt;-<span class="dv">1</span> <span class="co"># or can be a vector with one generation time for each species</span></span>
<span id="cb7-34"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-34" aria-hidden="true"></a></span>
<span id="cb7-35"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-35" aria-hidden="true"></a><span class="co">#extract information from the par file to the configuration table</span></span>
<span id="cb7-36"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-36" aria-hidden="true"></a>conf&lt;-<span class="kw">par_to_config</span>(bayessc_par_file,species,nloci,gen)</span>
<span id="cb7-37"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-37" aria-hidden="true"></a></span>
<span id="cb7-38"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-38" aria-hidden="true"></a><span class="co">#run BayeSSC with the configuration table</span></span>
<span id="cb7-39"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb7-39" aria-hidden="true"></a><span class="co">#simulatedobs&lt;-runbayeSSC_with_conf(BayeSSCallocation = path_to_bayessc,conf = conf,prefix = 'partemp',species.assignment = species.assignment,exp.time = exp.time,intern=TRUE)</span></span></code></pre></div>
</div>
<div id="iv-calculate-hyperstats-with-hbayssc" class="section level3">
<h3>iv) calculate hyperstats with hBaySSC</h3>
<p>We calculate hyperstats from the “observed” statistics. In our paper, this step was performed with <a href="https://github.com/UH-Bioinformatics/hBayeSSC">hBayeSSC</a>, but it depends on python 2.X which is no longer supported.Here we provide a conversion of the code to R.</p>
<ul>
<li>the obs file from previous step</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-1" aria-hidden="true"></a></span>
<span id="cb8-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-2" aria-hidden="true"></a><span class="co">#if you write the observed summary stats into a file</span></span>
<span id="cb8-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-3" aria-hidden="true"></a><span class="co">#obsfile&lt;-"temp_obs_file"</span></span>
<span id="cb8-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-4" aria-hidden="true"></a><span class="co">#simulatedobs&lt;-read.table(obsfile,sep="\t",header=T,stringsAsFactors = F)</span></span>
<span id="cb8-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-5" aria-hidden="true"></a></span>
<span id="cb8-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-6" aria-hidden="true"></a>hyperstat&lt;-<span class="kw">calculate_hyperstat</span>(simulatedobs)</span>
<span id="cb8-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-7" aria-hidden="true"></a><span class="co">#make a vector containing these hyperstats together the 'real' number of events </span></span>
<span id="cb8-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-8" aria-hidden="true"></a><span class="co">#and expansion time for each species </span></span>
<span id="cb8-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-9" aria-hidden="true"></a>a&lt;-<span class="kw">rep</span>(<span class="dv">0</span>,<span class="dv">3</span>)</span>
<span id="cb8-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-10" aria-hidden="true"></a>a[<span class="dv">1</span>]&lt;-<span class="st">'temp'</span> <span class="co">#or any name you want to give to this pseudo-observed dataset</span></span>
<span id="cb8-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-11" aria-hidden="true"></a>a[<span class="dv">2</span>]&lt;-<span class="kw">length</span>(exp.time) <span class="co">#true number of events</span></span>
<span id="cb8-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-12" aria-hidden="true"></a>a[<span class="dv">3</span>]&lt;-<span class="kw">length</span>(species) <span class="co">#total number of species</span></span>
<span id="cb8-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-13" aria-hidden="true"></a><span class="kw">names</span>(a)&lt;-<span class="kw">c</span>(<span class="st">"uid"</span>,<span class="st">"nevent"</span>,<span class="st">"nspecies"</span>)</span>
<span id="cb8-14"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-14" aria-hidden="true"></a>species.time&lt;-<span class="kw">species_exp_time</span>(<span class="dt">species.assignment =</span> species.assignment,<span class="dt">exp.time =</span> exp.time)</span>
<span id="cb8-15"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-15" aria-hidden="true"></a>a&lt;-<span class="kw">c</span>(a,species.time)</span>
<span id="cb8-16"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-16" aria-hidden="true"></a>hyperstat&lt;-<span class="kw">c</span>(a,hyperstat)</span>
<span id="cb8-17"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-17" aria-hidden="true"></a><span class="co">#write the hyperstat to a file</span></span>
<span id="cb8-18"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-18" aria-hidden="true"></a>outfile&lt;-<span class="st">"temp_hyperstat_file"</span></span>
<span id="cb8-19"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-19" aria-hidden="true"></a><span class="kw">cat</span>(<span class="kw">paste</span>(hyperstat,<span class="dt">sep=</span><span class="st">''</span>,<span class="dt">collapse =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>),<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>,<span class="dt">sep=</span><span class="st">''</span>,<span class="dt">file =</span> outfile)</span>
<span id="cb8-20"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb8-20" aria-hidden="true"></a>    </span></code></pre></div>
</div>
</div>
</div>
<div id="abc-simulation" class="section level1">
<h1><strong>2.ABC simulation</strong></h1>
<p>ABC simulation is very similar to generating test data, the only difference is that the event time should be autogenerated according to prior rather than specified by user.</p>
<p>For the prior on expansion event, hBayeSSC used a flat distribution where 1-nspecies expanded together are simulated with equal frequency. In the paper, we used a gamma distribution for the alpha parameter of the dirichelete process, similar to <a href="http://joaks1.github.io/PyMsBayes/">PyMsbayes</a>. User need to select the two parameters for gamma distribution: concentrationShape and consentrationScale. User can looked into the <a href="http://joaks1.github.io/PyMsBayes/tutorials/selecting-priors.html">prior selection in PyMsbayes</a> for how to select these two parameters</p>
<p>similar to step 1, user also need to provide:</p>
<ul>
<li>path to bayeSSC<br>
</li>
<li>the time range for possible expansion events<br>
</li>
<li>the buffer between any two events<br>
</li>
<li>number of POD to generate<br>
</li>
<li>Either
<ul>
<li>configuration file<br>
or if all species can use the same par file for BayeSSC<br>
</li>
<li>list of species<br>
</li>
<li>number of loci per species</li>
<li>generation time</li>
<li>.par file</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-1" aria-hidden="true"></a>path_to_bayessc=<span class="st">"BayeSSC.exe"</span></span>
<span id="cb9-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-2" aria-hidden="true"></a>concentrationShape=<span class="fl">20.0</span></span>
<span id="cb9-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-3" aria-hidden="true"></a>concentrationscale=<span class="fl">0.15</span></span>
<span id="cb9-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-4" aria-hidden="true"></a>time.range&lt;-<span class="kw">c</span>(<span class="dv">30000</span>,<span class="dv">50000</span>)</span>
<span id="cb9-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-5" aria-hidden="true"></a>buffer&lt;-<span class="dv">500</span></span>
<span id="cb9-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-6" aria-hidden="true"></a>npod&lt;-<span class="dv">10</span></span>
<span id="cb9-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-7" aria-hidden="true"></a></span>
<span id="cb9-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-8" aria-hidden="true"></a>conf&lt;-conf</span>
<span id="cb9-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-9" aria-hidden="true"></a></span>
<span id="cb9-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-10" aria-hidden="true"></a><span class="co">#running ABC simulation</span></span>
<span id="cb9-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-11" aria-hidden="true"></a><span class="co">#reference.table&lt;-ABC_simulation_with_conf(npod=npod,conf=conf,time.range=time.range,buffer=buffer,concentrationscale=concentrationscale,concentrationShape=concentrationShape,BayeSSCallocation=path_to_bayessc,prefix='temp',do.parallel=2,write.reference.file = F)</span></span>
<span id="cb9-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-12" aria-hidden="true"></a></span>
<span id="cb9-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-13" aria-hidden="true"></a><span class="co">#the package can with a toy example of the reference.table</span></span>
<span id="cb9-14"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-14" aria-hidden="true"></a><span class="kw">colnames</span>(reference.table)</span>
<span id="cb9-15"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-15" aria-hidden="true"></a><span class="co">#&gt;  [1] "uid"                "nevent"             "nspecies"          </span></span>
<span id="cb9-16"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-16" aria-hidden="true"></a><span class="co">#&gt;  [4] "1"                  "10"                 "2"                 </span></span>
<span id="cb9-17"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-17" aria-hidden="true"></a><span class="co">#&gt;  [7] "3"                  "4"                  "5"                 </span></span>
<span id="cb9-18"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-18" aria-hidden="true"></a><span class="co">#&gt; [10] "6"                  "7"                  "8"                 </span></span>
<span id="cb9-19"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-19" aria-hidden="true"></a><span class="co">#&gt; [13] "9"                  "haptypes_Mean"      "haptypes_Variance" </span></span>
<span id="cb9-20"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-20" aria-hidden="true"></a><span class="co">#&gt; [16] "haptypes_Skewness"  "haptypes_Kurtosis"  "hapdiv_Mean"       </span></span>
<span id="cb9-21"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-21" aria-hidden="true"></a><span class="co">#&gt; [19] "hapdiv_Variance"    "hapdiv_Skewness"    "hapdiv_Kurtosis"   </span></span>
<span id="cb9-22"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-22" aria-hidden="true"></a><span class="co">#&gt; [22] "nucdiv_Mean"        "nucdiv_Variance"    "nucdiv_Skewness"   </span></span>
<span id="cb9-23"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-23" aria-hidden="true"></a><span class="co">#&gt; [25] "nucdiv_Kurtosis"    "tajimasD_Mean"      "tajimasD_Variance" </span></span>
<span id="cb9-24"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-24" aria-hidden="true"></a><span class="co">#&gt; [28] "tajimasD_Skewness"  "tajimasD_Kurtosis"  "fusf_Mean"         </span></span>
<span id="cb9-25"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-25" aria-hidden="true"></a><span class="co">#&gt; [31] "fusf_Variance"      "fusf_Skewness"      "fusf_Kurtosis"     </span></span>
<span id="cb9-26"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-26" aria-hidden="true"></a><span class="co">#&gt; [34] "pairdiffs_Mean"     "pairdiffs_Variance" "pairdiffs_Skewness"</span></span>
<span id="cb9-27"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-27" aria-hidden="true"></a><span class="co">#&gt; [37] "pairdiffs_Kurtosis" "segSites_Mean"      "segSites_Variance" </span></span>
<span id="cb9-28"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-28" aria-hidden="true"></a><span class="co">#&gt; [40] "segSites_Skewness"  "segSites_Kurtosis"</span></span>
<span id="cb9-29"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-29" aria-hidden="true"></a><span class="kw">head</span>(reference.table[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</span>
<span id="cb9-30"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-30" aria-hidden="true"></a><span class="co">#&gt;      uid     nevent nspecies 1                  10                </span></span>
<span id="cb9-31"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-31" aria-hidden="true"></a><span class="co">#&gt; [1,] "temp1" "9"    "10"     "39359.1877259314" "36782.8306816535"</span></span>
<span id="cb9-32"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-32" aria-hidden="true"></a><span class="co">#&gt; [2,] "temp2" "6"    "10"     "36854.4358434156" "36854.4358434156"</span></span>
<span id="cb9-33"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-33" aria-hidden="true"></a><span class="co">#&gt; [3,] "temp3" "6"    "10"     "30039.6010745317" "30039.6010745317"</span></span>
<span id="cb9-34"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-34" aria-hidden="true"></a><span class="co">#&gt; [4,] "temp4" "3"    "10"     "32991.7780589312" "45625.935561955" </span></span>
<span id="cb9-35"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-35" aria-hidden="true"></a><span class="co">#&gt; [5,] "temp5" "5"    "10"     "34564.4979132339" "41061.2173192203"</span></span>
<span id="cb9-36"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb9-36" aria-hidden="true"></a><span class="co">#&gt; [6,] "temp6" "5"    "10"     "36073.3391717076" "34834.24140516"</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-1" aria-hidden="true"></a>path_to_bayessc=<span class="st">"BayeSSC.exe"</span></span>
<span id="cb10-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-2" aria-hidden="true"></a>concentrationShape=<span class="fl">20.0</span></span>
<span id="cb10-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-3" aria-hidden="true"></a>concentrationscale=<span class="fl">0.15</span></span>
<span id="cb10-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-4" aria-hidden="true"></a>time.range&lt;-<span class="kw">c</span>(<span class="dv">30000</span>,<span class="dv">50000</span>)</span>
<span id="cb10-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-5" aria-hidden="true"></a>buffer&lt;-<span class="dv">500</span></span>
<span id="cb10-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-6" aria-hidden="true"></a>npod&lt;-<span class="dv">100</span></span>
<span id="cb10-7"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-7" aria-hidden="true"></a></span>
<span id="cb10-8"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-8" aria-hidden="true"></a><span class="co"># you can write the template out to a file and edit it with any text editor</span></span>
<span id="cb10-9"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-9" aria-hidden="true"></a><span class="kw">cat</span>(<span class="kw">paste</span>(templateparfile,<span class="dt">sep =</span> <span class="st">''</span>,<span class="dt">collapse =</span> <span class="st">'</span><span class="ch">\n</span><span class="st">'</span>),<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>,<span class="dt">sep=</span><span class="st">''</span>,<span class="dt">file=</span><span class="st">"test.par"</span>)</span>
<span id="cb10-10"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-10" aria-hidden="true"></a>bayessc_par_file&lt;-<span class="st">'test.par'</span></span>
<span id="cb10-11"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-11" aria-hidden="true"></a>nloci&lt;-<span class="dv">10</span></span>
<span id="cb10-12"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-12" aria-hidden="true"></a>species&lt;-<span class="dv">1</span><span class="op">:</span><span class="dv">10</span></span>
<span id="cb10-13"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-13" aria-hidden="true"></a>gen&lt;-<span class="dv">1</span> <span class="co"># or can be a vector with one generation time for each species</span></span>
<span id="cb10-14"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-14" aria-hidden="true"></a></span>
<span id="cb10-15"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-15" aria-hidden="true"></a>conf&lt;-<span class="kw">par_to_config</span>(bayessc_par_file,species,nloci,gen)</span>
<span id="cb10-16"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-16" aria-hidden="true"></a><span class="co">#reference.table&lt;-ABC_simulation_with_conf(npod=npod,conf=conf,time.range=time.range,buffer=buffer,concentrationscale=concentrationscale,concentrationShape=concentrationShape,BayeSSCallocation=path_to_bayessc,prefix='temp',do.parallel=2)</span></span>
<span id="cb10-17"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-17" aria-hidden="true"></a></span>
<span id="cb10-18"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb10-18" aria-hidden="true"></a><span class="co">#run BayeSCC for each species</span></span></code></pre></div>
</div>
<div id="generating-inference-for-test-data-run-msreject" class="section level1">
<h1><strong>3.Generating Inference for Test Data (run MsReject)</strong></h1>
<p>For installing/compiling msReject see <a href="http://msbayes.sourceforge.net/">msbayes webpage</a>. User need to provide * the “observed” hyperstats<br>
* the simulated reference data table<br>
* the path to msReject<br>
* sampling tolerance in line 16<br>
* prefix to the posterior file</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb11-1" aria-hidden="true"></a>path_to_msreject&lt;-<span class="st">"./msReject"</span></span>
<span id="cb11-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb11-2" aria-hidden="true"></a>samplingtolerance&lt;-<span class="fl">0.5</span> <span class="co">#just toy example, should be much smaller for real runs</span></span>
<span id="cb11-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb11-3" aria-hidden="true"></a>prefix&lt;-<span class="st">'temp'</span></span>
<span id="cb11-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb11-4" aria-hidden="true"></a></span>
<span id="cb11-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb11-5" aria-hidden="true"></a><span class="co">#posterior&lt;-run_msreject(hyperstat=hyperstat,reference.table=reference.table,MsRejectallocation=path_to_msreject,samplingtolerance=samplingtolerance,prefix=prefix)</span></span></code></pre></div>
<p>With this posterior after msreject, we can use the VGAM, abc and locfit package to do the final acceptance and parameter estimation, some code are on the <a href="https://github.com/UH-Bioinformatics/hBayeSSC#msreject-module">hBayeSSC webpage</a>. Here we can just look at the mode of event number in the posterior samples.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb12-1" aria-hidden="true"></a>getmode &lt;-<span class="st"> </span><span class="cf">function</span>(v) {</span>
<span id="cb12-2"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb12-2" aria-hidden="true"></a>  uniqv &lt;-<span class="st"> </span><span class="kw">unique</span>(v)</span>
<span id="cb12-3"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb12-3" aria-hidden="true"></a>  uniqv[<span class="kw">which.max</span>(<span class="kw">tabulate</span>(<span class="kw">match</span>(v, uniqv)))]</span>
<span id="cb12-4"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb12-4" aria-hidden="true"></a>}</span>
<span id="cb12-5"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb12-5" aria-hidden="true"></a><span class="kw">getmode</span>(posterior<span class="op">$</span>nevent)</span>
<span id="cb12-6"><a href="file:///C:/Users/huang/AppData/Local/Temp/RtmpqCWElF/preview-24f83ce215ff.dir/examplerun.html#cb12-6" aria-hidden="true"></a><span class="co">#&gt; [1] 4</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>



</body></html>